packages/utils/src/bulk-query-patients.ts

```diff
@@ -0,0 +1,124 @@
+const maxDocQueryAttemts = 3;
+
+// csv stuff
+const cxName = getEnvVarOrFail("CX_NAME");
+const csvHeader =
+  "patientId,firstName,lastName,state,queryAttemptCount,docCount,fhirResourceCount,fhirResourceDetails,status\n";
+const curDateTime = new Date();
+const csvName = `./${replaceAll(cxName, " ", "").trim()}-DocQuery-${curDateTime.toISOString()}.csv`;
+
+function replaceAll(string: string, search: string, replace: string): string {
+  return string.split(search).join(replace);
+}
+
+function initCsv() {
+  fs.writeFileSync(csvName, csvHeader);
+}
+
+function sleep(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function queryDocsForPatient(patient: Patient) {
+  let docQueryAttempts = 0;
+  let docCount = 0;
+  let fhirResourceCount = 0;
+  let status: "completed" | "docs-not-found" = "docs-not-found";
+  let queryComplete = false;
+  let fhirResourceCounts: ConsolidatedCountResponse = {
+    filter: { resources: "" },
+    resources: {},
+    total: 0,
+  };
+  while (docQueryAttempts < maxDocQueryAttemts) {
+    console.log(`>>> Starting doc query for patient ${patient.id}...`);
+    await metriportAPI.startDocumentQuery(patient.id, patient.facilityIds[0]);
+    // add a bit of jitter to the requests
+    await sleep(200 + Math.random() * patientChunkDelayJitterMs);
+    const queryStartTime = Date.now();
+    while (Date.now() - queryStartTime < maxQueryDurationMs) {
+      const docQueryStatus = await metriportAPI.getDocumentQueryStatus(patient.id);
+      // ensure at least 1 doc was downloaded, and both the download + conversion have completed
+      if (
+        docQueryStatus.download &&
+        docQueryStatus.download.status === "completed" &&
+        docQueryStatus.download.successful &&
+        docQueryStatus.download.successful > 0 &&
+        (!docQueryStatus.convert ||
+          (docQueryStatus.convert && docQueryStatus.convert.status === "completed"))
+      ) {
+        queryComplete = true;
+        docCount = docQueryStatus.download.successful;
+        break;
+      }
+      await sleep(queryPollDurationMs);
+    }
+    docQueryAttempts++;
+    if (queryComplete) break;
+    console.log(`>>> Didn't find docs for patient ${patient.id} on attmept ${docQueryAttempts}...`);
+  }
+  if (queryComplete) {
+    status = "completed";
+    // get count of resulting FHIR resources
+    fhirResourceCounts = await metriportAPI.countPatientConsolidated(patient.id);
+    for (const val of Object.values(fhirResourceCounts.resources)) {
+      fhirResourceCount += val;
+    }
+  }
+
+  // write line to results csv
+  const state = Array.isArray(patient.address) ? patient.address[0].state : patient.address.state;
+  fs.appendFileSync(
+    csvName,
+    `${patient.id},${patient.firstName},${
+      patient.lastName
+    },${state},${docQueryAttempts},${docCount},${fhirResourceCounts},${replaceAll(
+      JSON.stringify(fhirResourceCount),
+      ",",
+      " "
+    )},${status}\n`
+  );
```

Comment:
NABD, but ideally would prep these into individual variables so we're not trying to understand how each gets build/purpose, as well as how they get arranged